Initialize Security Disclosure Processes

ShubhamPalriwala opened the pull request:
## Related issue
Closes #2055 

## What type of PR is this
> /kind feature

## Proposed Changes
Initialise `security.md` which introduces
- When to and when not to report a vulnerability
- Product Security Team
- Public and Private Security Disclosure Processes  

Inspiration has been taken from various ongoing and graduated CNCF projects to maintain uniformity.

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [] I have added tests that prove my fix is effective or that my feature works.
- [] My PR contains new or altered behavior to Kyverno and
  - [] I have added or changed [the documentation](https://github.com/kyverno/website) myself in an existing PR and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->
  - [] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the doc update and the link is:
  <!-- Uncomment to link to the issue -->
  <!-- https://github.com/kyverno/website/issues/1 -->
  - [] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.

MarcelMue commented on SECURITY.md:
```diff
@@ -0,0 +1,116 @@
+## Reporting a Vulnerability
+
+### When you should?
+- You think you discovered a potential security vulnerability in Kyverno
```
Just a nit before deeper review:
Please make sure to have a full stop at the end of each sentence & make sure empty lines are consistent (sometimes it's one or two empty lines between headings).

ShubhamPalriwala replied:
Thanks for the suggestion. Didn't notice it. Have made the empty lines consistent and added full stops wherever necessary

JimBugwadia commented on SECURITY.md:
Not exactly sure what this means....can you please elaborate on the intent?

ShubhamPalriwala replied:
This is in general with respect to the security of the project. If anybody wants to contribute in the security aspect of Kyverno, they can help to do so in technical as well as administrative ways.

JimBugwadia commented on SECURITY.md:
Can you please share a reference for projects that use this.

ShubhamPalriwala replied:
Yeah. Some of them are [Opa](https://github.com/open-policy-agent/opa/blob/main/SECURITY.md), [Envoy Proxy](https://github.com/envoyproxy/envoy/blob/main/SECURITY.md). They mainly use it for their private distributors.

JimBugwadia reviewed (commented):
Shubham, in general this looks good but is far more elaborate than https://github.com/cncf/tag-security/blob/main/project-resources/templates/SECURITY.md. 

Since the project is relatively small, may make since to keep this minimal. What do you think?

Also, can you please add a `Credits` section and reference the projects we have leveraged? Something like this: https://github.com/sigstore/community/blob/main/SECURITY.md#credit?

ShubhamPalriwala commented:
Yeah, this is elaborated more than the one mentioned in the [tag-security template](https://github.com/cncf/tag-security/blob/main/project-resources/templates/SECURITY.md). This might be because we have covered the `disclosure timeline` in detail and have defined the embargo policy too as mentioned in the template.

In my opinion, if we continue with the Disclose timelines from now itself then it would be a great practice to maintain and fix vulnerabilities according to the process. This will help us in the long run.

However, as you suggested to keep this minimal, we can cut down on the detailed disclosure process and the Product Security Team and make this more of a crisp `How to report a vulnerability` for our contributors.
Let me know what do you think would be the ideal approach to go ahead with this.

And with respect to the credits, will definitely add them once the content and policies are finalised.

JimBugwadia commented:
Hi @ShubhamPalriwala yes, lets start with the minimal required processes and expand. We can keep the current focus to processes for reporting security issues, as you suggested.  Please update accordingly, and we can review and finalize.

ShubhamPalriwala commented:
Okay! I have updated `security.md` by keeping the [tag-security template](https://github.com/cncf/tag-security/blob/main/project-resources/templates/SECURITY.md) as the base.

### Changes I've made from the base file:
- Added a `When to report a vulnerability` and `When not to report a vulnerability` section.
This should be a part of this file as it clears the confusion for a user and avoids unwanted emails.
- Removed the option to fill a `web form` to submit a vulnerability as we do not have it currently.

### We need to decide on the following things:
- Do we need a web form 
An issue is already opened on [kverno/website#225](https://github.com/kyverno/website/issues/225) but are we planning to add a form there?
- Since we are handling every security aspect privately, I don't think so that we need an Embargo policy.
- Either we can have a fixed disclosure timeline of **X days** or not mention it so that we can be flexible based on the vulnerability.
Would like to know your opinion on these.

JimBugwadia commented:
Thanks @ShubhamPalriwala!

> Do we need a web form
> An issue is already opened on kverno/website#225 but are we planning to add a form there?

No, we can start with the email list and later add a web form if needed.

> Since we are handling every security aspect privately, I don't think so that we need an Embargo policy.

Agreed.. Lets remove for now.

> Either we can have a fixed disclosure timeline of X days or not mention it so that we can be flexible based on the vulnerability.
> Would like to know your opinion on these.

Can we specify a fixed initial response time, and then a range for the fix based on complexity and severity?

JimBugwadia commented:
@ShubhamPalriwala - couple of minor changes, and then we can merge!

Please remove this line "This project follows a disclosure timeline. Refer to our embargo policy for more information."

For the versions table, seems like we should replace with a general statement on versions supported, so we do not need to update this table each time a new version is released. What do you think? Can we state something like, "A maximum of three prior minor versions will be supported for security updates."?

ShubhamPalriwala commented:
Thanks Jim, I have made the changes and will be pushing them all together as soon as the below queries get solved!

> Can we specify a fixed initial response time, and then a range for the fix based on complexity and severity?

Okay! So can we set a 7-day time limit to acknowledge the vulnerability report that we receive?
And should we specify a range of like 7-28 days based on the complexity and the severity of the vulnerability or rather not mention the time duration here at all.

> For the versions table, seems like we should replace with a general statement on versions supported, so we do not need to update this table each time a new version is released. What do you think? Can we state something like, "A maximum of three prior minor versions will be supported for security updates."?

Okay this is more clear and to the point in my opinion so let's go ahead with this. So can we set "3 prior minor versions" here?

JimBugwadia commented:
> Okay! So can we set a 7-day time limit to acknowledge the vulnerability report that we receive?
And should we specify a range of like 7-28 days based on the complexity and the severity of the vulnerability or rather not mention the time duration here at all.

Ok, we can state with something like this?

"The project security team will send an initial response to the disclosure in 3-5 days. Once the vulnerability and fix are confirmed the team will plan to release the fix in 7 to 28 days, based on the severity and complexity."

> Okay this is more clear and to the point in my opinion so let's go ahead with this. So can we set "3 prior minor versions" here?

Here is the suggested text:

(inspired by https://kubernetes.io/releases/version-skew-policy/)

"Kyverno versions are expressed as x.y.z, where x is the major version, y is the minor version, and z is the patch version, following [Semantic Versioning](https://semver.org/) terminology. Security fixes, may be backported to those three release branches, depending on severity and feasibility. Patch releases are cut from those branches periodically, plus additional urgent releases, when required."

ShubhamPalriwala commented:
Thanks for the suggestions @JimBugwadia, I have made the changes as suggested. Modified the **Supported Versions** content a little bit from what you suggested. Please have a look and let me know if there are any changes.

JimBugwadia commented on SECURITY.md:
```diff
@@ -0,0 +1,34 @@
+- y is the minor version
+- and z is the patch version
+
+following [Semantic Versioning](https://semver.org/) terminology. Security fixes, may be backported to the three most recent minor releases, depending on severity and feasibility. Patch releases are cut from those branches periodically, plus additional urgent releases, when required.
```
Minor change to improve readability:

Kyverno versions follow  [Semantic Versioning](https://semver.org/) terminology and are expressed as x.y.z:
- where x is the major version
- y is the minor version
- and z is the patch version

ShubhamPalriwala replied:
Done! Thank you for the suggestion :)
